Update Apex writing and testing (#1967)
forcedotcom/salesforcedx-vscode: docs/_articles/en/apex/writing.md

lang: en
---

> Use the Apex extension for VS Code to access code-editing features such as code completion, go to definition, view outlines of Apex classes and triggers, refactoring, and find syntactic errors in your code.

Reviewer: I would recommend keeping this line: The Apex extension is powered by the [Apex Language Server](./en/apex/language-server). This is the only reference to the Apex Language Server in this doc.

Author: Deleting the Apex Language Server topic based on Nate and Stephanie's suggestion. So, deleted the reference.

## Code-Completion

> Apex extension provides context-sensitive suggestions when you’re working in an Apex class or trigger. As you type, the auto-completion list of members such as, methods or variables pop up. The auto-completion list also displays the documentation for the suggestions. Use these keys:

Reviewer: ```suggestion
Apex extension provides context-sensitive suggestions when you’re working in an Apex class or trigger. As you type, auto-completion will list members such as methods or variables. The auto-completion list also displays the documentation for the suggestions. Use these keys:
```

Author: Done

> - Ctrl+space to view see code-completion suggestions

Reviewer: ```suggestion
- Ctrl+space to view code-completion suggestions
```

Author: Done

## Go To Definitions

Apex extension provides go to definition support for user-defined Apex such as classes and methods, standard objects, and custom objects.

- To preview a definition, press and hold Ctrl (Windows or Linux) or Cmd (macOS) and hover over the item whose definition you want to see.
- To view a definition, right-click the item and select **Peek Definition**, or press Alt+F12.
- To jump to the location of a definition, right-click the item and select **Go to Definition**, or press F12.

![Previewing, viewing, and jumping to a definition](../../images/apex_go_to_definition.gif)

## Find All References

You can find all references to user-defined Apex such as Classes, Class variables, Enums, Interfaces, Methods, and Properties. To find references, select an item and press Shift+F12; Or right-click the item and select **Find All References**. The reference results are displayed in the left pane of the editor window.

## Outline View

The Apex outline view shows the structure of the Apex class or trigger that’s open in the editor.

> - To view the list of the symbols in the file, press Ctrl+Shift+O (Windows or Linux) or Cmd+Shift+O (macOS).

Reviewer: I know this shortcut was listed previously, but is this right? Cmd+Shift+0 doesn't bring anything up for me in an Apex class. I have not used this shortcut before, so just checking. @brpowell do you happen to know?

Reviewer 2: I believe that's an 'O' not '0' :), the former works.

Reviewer: @brpowell perfect - Thanks :)

- To go to one of the symbols, select it from the list.

![Outline view, showing the symbols in an Apex class](../../images/apex_outline.png)

The Explorer in the Side Bar also provides an Outline view to assist you while working on your project.

## Syntax Errors

Any syntax errors such as a missing semi-colon or a bracket are marked with a red squiggly line in the editor. The Problems panel also lists the syntax errors. To go to the source file with the syntax error, double-click the error.

![Problems view, showing a missing semicolon in an Apex class](../../images/apex_problems.png)

## Quick Fix

When you reference a method that isn’t declared in your source, use the Quick Fix widget to automatically declare the method.

> To invoke the Declare Missing Methods quick fix in VS Code, use one these ways:

Reviewer: ```suggestion
Declare Missing Methods quick fix can be invoked in the following ways:
```
